_posts/2021-03-23-keccak-optimizer-bug.md

```diff
@@ -0,0 +1,118 @@
+      // instead of 
+      // 0xe2b9f9f9430b05bfa9a3abd3bac9a181434d23a707ef1cde8bd25d30203538d8
+      b := keccak256(0, 23)
+    }
+  }
+}
+```
+
+In the above example, both the Keccak-256 hashes can be computed at compile time. Here are two other
+cases where the hashes cannot be computed at compile time, yet the optimizer considered them equal
+and thus modified the semantics:
+
+```solidity
+contract C {
+  function bug(string memory s) public returns (bool ret) {
+    assembly {
+      let a := keccak256(s, 32)
+      let b := keccak256(s, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+```solidity
+contract C {
+  function bug() public view returns (bool ret) {
+    assembly {
+      let x := calldataload(0)
+      mstore(0, x)
+      mstore(0x20, x)
+      let a := keccak256(0, 4)
+      // even though the memory location is different,
+      // the 32-byte content is the same.
+      let b := keccak256(0x20, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+Specifically, `keccak256(mpos1, length1)` and `keccak256(mpos2, length2)` in some cases were
+considered equal if `length1` and `length2`, when rounded up to nearest multiple of 32 were the
+same, and when the memory contents at `mpos1` and `mpos2` can be deduced to be equal.
+
+You maybe **affected** if you compute multiple Keccak-256 hashes of the same content, but with
+different length inside inline assembly, and enabled the optimizer. 
+
+### Unaffected Cases
+
+You are unaffected in the following situations:
+
+- If your code uses `keccak256` with a length that is not known at compile time.
+- If your code only uses `keccak256` with length that is always a multiple of 32. In particular, the
+  Solidity compiler always uses multiples of 32 for storage slot computation.
+- If pairs of `keccak256` hashes are separated by certain instructions such as:
+
+  - An instruction that breaks the control flow, e.g., `jumpi`, `jump` etc. In high level code, this
+     would be function calls, if statements, loops, etc.
+  - An instruction that writes to memory other than `mstore`, e.g., `call`, `returndatacopy`, etc..
+
+### Severity
+
+Since most usages of `keccak256` falls under one of the unaffected cases mentioned in the previous
+section, we assigned the bug a severity level "medium".
+
+### Fuzzing
+
+The bug was discovered by differentially testing the current code-generation and the upcoming Yul
+based code-generation with fuzzer generated input. Essentially, the example in
+[solidity#11132](https://github.com/ethereum/solidity/issues/11131) produced different results, when
+compiled using different code-generation paths; this is because the bytecode optimizer is only
+activated in the old code-generation, and the Yul based codegen has a more powerful Yul-optimizer.
```

Comment:
```suggestion
activated in the old code-generation, while the Yul based codegen uses the Yul-optimizer that does not contain the bug.
```

Reply:
I added it with a small rewording. There is a PR by @axic to enable bytecode optimizer in the Yul codegen.